### ISSUE TITLE
Feature request: "export" functionality

### ISSUE THREAD

**Issue by carlosgalvezp** [martis42/depend_on_what_you_use#138]
Hi!

I was wondering if there are plans for adding some sort of "export" functionality to DWYU, similar to the equivalent functionality in IWYU. We currently have issues on googletests, because DWYU forces us to duplicate the deps:

```
        "@googletest//:gtest",
        "@googletest//:gtest_main",
```
It would be great if we could tell DWYU that `gtest_main` _exports_ `gtest`, so it's fine to include only `gtest_main`.

Thanks!

**Comment by martis42** [comment]
Hi @carlosgalvezp,

strictly speaking it is correct to provide both `@googletest//:gtest` and `@googletest//:gtest_main` as dependency to a `cc_test` target. `gtest_main` is not providing itself the gtest headers, but merely depends on the headers to compile the gtest main function, see [here](https://github.com/google/googletest/blob/main/BUILD.bazel#L154). Thus, depending on `gtest_main` and using `#include <gtest/gtest.h>` is actually a case on relying on headers from a transitive dependency, which DWYU wants to prevent. Theoretically, the gtest authors could at any point decide to utilize `implementatoin_deps` to compile `gtest_main` and make the gtest headers unavailable to the `gtest_main` users (although this happening is quite unlikely).
In other words, your gtest dependency is not duplicated, but you depend on 2 distinguished targets which provide you with different things (the gtest library itself and the convenience main function to not have to provide your own).

All that said, I understand that theory and what is practical are not always the same.
I will think about he feature request. If I see a clean way to implement this I will do.
But at the same time I can't promise this will be resolved soon, as some other things have higher priority to me right now.

**Comment by carlosgalvezp** [comment]
Thanks for the reply @martis42 ! I fully agree with your comment and in most cases I agree with the philosophy of *WYU. 

In this particular it does not seem that the googletest authors intended users to put both `gtest` and `gtest_main` as dependencies, see here for a sample use case:

https://github.com/google/googletest/blob/main/BUILD.bazel#L203

We are thus trying to follow the intended workflow they have setup, but need to currently duplicate the includes due to DWYU.

I totally understand they are more urgent things to fix, no rush!

**Comment by carlosgalvezp** [comment]
Another use case that we come across every now and then is:

```
cc_library(
  name = "foo_headers",
  hdrs = [...],
)

cc_library(
  name = "foo_x86",
  srcs = [...],
  deps = [":foo_headers"],
)

cc_library(
  name = "foo_aarch64",
  srcs = [...],
  deps = [":foo_headers"],
)
```

It can certainly be solved in other ways like `select`'ing on the headers in a single `foo` library, but at some point it becomes too messy and it's better to create 2 independent libraries.

**Comment by martis42** [comment]
Thanks for outlining this use case.

As you said yourself, one can model this in a DWYU compliant way. However, the multiple ways to model this have their own pros and cons. DWYU should not be so opinionated that it enforces a specific pattern.

I still don't know when I will have time to work on this. But with this latest example I am fully accepting this feature request (aka promoting this from "nice to have" to "will definitely do if there is a feasible implementation").

**Comment by carlosgalvezp** [comment]
Sounds great, thank you! Let me know if I can help in any way :)

**Comment by martis42** [comment]
@carlosgalvezp 

This was more complex than anticipated. Still, release [0.0.13](https://github.com/martis42/depend_on_what_you_use/releases/tag/0.0.13) should resolve your request. Please have a look.

I hope the new feature / API does work for you.

**Comment by carlosgalvezp** [comment]
Thanks a lot for the quick help! Will test it right away :)
